crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -543,49 +570,147 @@ impl Object {
+
+                let lhs: i64 = conn
+                    .first(move || {
+                        snapshot::objects_snapshot
+                            .select(snapshot::checkpoint_sequence_number)
+                            .order(snapshot::checkpoint_sequence_number.desc())
+                    })
+                    .optional()?
+                    .unwrap_or_default();
+
+                if let Some(checkpoint_sequence_number) = checkpoint_sequence_number {
+                    if checkpoint_sequence_number > rhs as u64
+                        || checkpoint_sequence_number < lhs as u64
+                    {
+                        return Ok::<_, diesel::result::Error>(
+                            HistoricalObjectPaginationResult::Error(
+                                HistoricalObjectPaginationError::OutsideAvailableRange(
+                                    checkpoint_sequence_number,
+                                    lhs as u64,
+                                    rhs as u64,
+                                ),
+                            ),
                         );
                     }
+                    rhs = checkpoint_sequence_number as i64;
+                }
 
-                    for ObjectKey { object_id, version } in filter.object_keys.iter().flatten() {
-                        query = query.or_filter(
-                            dsl::object_id
-                                .eq(object_id.into_vec())
-                                .and(dsl::object_version.eq(*version as i64)),
+                let result =
+                    page.paginate_raw_query::<StoredHistoryObject, _>(conn, move || {
+                        let start_cp = lhs;
+                        let end_cp = rhs;
+
+                        let mut snapshot_query = RawSqlQuery {
+                            sql: "SELECT * FROM objects_snapshot".to_string(),
+                            alias: "objects_snapshot".to_string(),
+                            has_where_clause: false,
+                        };
+                        let mut history_query = RawSqlQuery {
+                            sql: "SELECT * FROM objects_history".to_string(),
+                            alias: "objects_history".to_string(),
+                            has_where_clause: false,
+                        };
+                        let checkpoint_bound = format!(
+                            "WHERE checkpoint_sequence_number BETWEEN {} AND {}",
+                            start_cp, end_cp
                         );
-                    }
 
-                    if let Some(type_) = &filter.type_ {
-                        query = query.filter(dsl::object_type.is_not_null());
-                        query = type_.apply(query, dsl::object_type.assume_not_null());
-                    }
+                        Object::raw_object_filter(&mut snapshot_query, &filter);
+                        Object::raw_object_filter(&mut history_query, &filter);
```

Comment:
Yes, it looks like we would have to do the filtering twice, once to build the sql query, and again to bind the parameters. There is no way to access the dsl query's bindings, so it would have to be done manually.